state/spaces.go

```diff
@@ -529,13 +530,57 @@ func (st *State) SaveProviderSpaces(providerSpaces []network.SpaceInfo) error {
 			}
 
 			spaceNames.Add(spaceName)
-			spaceId = space.Id()
+			spaceID = space.Id()
+
+			// To ensure that we can remove spaces, we backfill the new spaces
+			// onto the modelSpaceMap.
+			modelSpaceMap[space.ProviderId()] = space
 		}
 
-		err = st.SaveProviderSubnets(spaceInfo.Subnets, spaceId)
+		err = st.SaveProviderSubnets(spaceInfo.Subnets, spaceID)
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		updatedSpaces.Add(spaceInfo.ProviderId)
+	}
+
+	// Workout the difference between all the current spaces vs what was
+	// actually changed.
+	allStateSpaces := network.MakeSubnetSet()
+	for providerID := range modelSpaceMap {
+		allStateSpaces.Add(providerID)
+	}
+
+	remnantSpaces := allStateSpaces.Difference(updatedSpaces)
+	for _, providerID := range remnantSpaces.SortedValues() {
+		// If the space is not in state or the name is not in space names, then
+		// we can ignore it.
+		space, ok := modelSpaceMap[providerID]
+		if !ok || space.Name() == network.AlphaSpaceName {
+			continue
+		}
+
```

Comment:
I think somewhere about here, we want to check that the space to be deleted is not present in constraints or endpoint bindings.